File: backend/src/main/java/com/intellihealth/service/WorkoutService.java
package com.intellihealth.service;

import com.intellihealth.entity.WorkoutPlan;
import com.intellihealth.entity.Exercise;
import com.intellihealth.entity.ExerciseCategoryEntity;
import com.intellihealth.entity.User;
import com.intellihealth.repository.WorkoutPlanRepository;
import com.intellihealth.repository.ExerciseRepository;
import com.intellihealth.repository.ExerciseCategoryRepository;
import com.intellihealth.dto.WorkoutPlanCreateRequest;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.Arrays;
import java.util.List;
import java.util.Optional;

@Service
@RequiredArgsConstructor
@Slf4j
public class WorkoutService {

    private final WorkoutPlanRepository workoutPlanRepository;
    private final ExerciseRepository exerciseRepository;
    private final ExerciseCategoryRepository exerciseCategoryRepository;
    private final UserService userService;

    @Transactional
    public WorkoutPlan createWorkoutPlan(WorkoutPlanCreateRequest request) {
        User currentUser = userService.getCurrentUser();
        
        Exercise exercise = exerciseRepository.findById(request.getExerciseId())
                .orElseThrow(() -> new RuntimeException("Exercise not found with id: " + request.getExerciseId()));
        
        WorkoutPlan workoutPlan = WorkoutPlan.builder()
                .user(currentUser)
                .exercise(exercise)
                .exerciseType(request.getExerciseType())
                .title(request.getTitle() != null ? request.getTitle() : exercise.getName() + " Workout")
                .description(request.getDescription() != null ? request.getDescription() : exercise.getDescription())
                .workoutDate(request.getWorkoutDate())
                .startTime(request.getStartTime())
                .endTime(request.getEndTime())
                .comments(request.getComments())
                .workoutPhotoUrl(request.getWorkoutPhotoUrl())
                .activeStatus(WorkoutPlan.ActiveStatus.ACTIVE)
                .build();
                
        return workoutPlanRepository.save(workoutPlan);
    }

    @Transactional
    public WorkoutPlan createWorkoutPlan(WorkoutPlan workoutPlan) {
        User currentUser = userService.getCurrentUser();
        workoutPlan.setUser(currentUser);
        return workoutPlanRepository.save(workoutPlan);
    }

    @Transactional(readOnly = true)
    public List<Exercise> getExercisesByCategory(String category) {
        try {
            log.info("Fetching exercises for category: '{}'", category);
            // Check if category exists first
            Optional<ExerciseCategoryEntity> categoryEntity = exerciseCategoryRepository.findByName(category);
            if (categoryEntity.isEmpty()) {
                log.warn("Category '{}' not found in database", category);
                return List.of();
            }
            // Use category name query instead of entity to avoid relationship issues
            List<Exercise> exercises = exerciseRepository.findByCategoryName(category);
            log.info("Found {} exercises for category '{}'", exercises.size(), category);
            return exercises;
        } catch (Exception e) {
            log.error("Error fetching exercises for category '{}': {}", category, e.getMessage(), e);
            return List.of();
        }
    }

    @Transactional(readOnly = true)
    public List<String> getExerciseTypes() {
        log.info("Fetching available exercise types");
        try {
            List<ExerciseCategoryEntity> categories = exerciseCategoryRepository.findAll();
            List<String> types = categories.stream()
                    .map(ExerciseCategoryEntity::getName)
                    .toList();
            log.info("Available exercise types: {}", types);
            return types;
        } catch (Exception e) {
            log.error("Error fetching exercise types: {}", e.getMessage());
            return Arrays.asList("AEROBIC", "STRENGTH", "FLEXIBILITY", "BALANCE");
        }
    }

    @Transactional(readOnly = true)
    public List<WorkoutPlan> getTodayWorkoutPlans() {
        User currentUser = userService.getCurrentUser();
        java.time.LocalDate today = java.time.LocalDate.now();
        log.info("Getting today's workouts for user {} (ID: {}) on date {}", 
                 currentUser.getEmail(), currentUser.getId(), today);
        List<WorkoutPlan> plans = workoutPlanRepository.findByUserAndWorkoutDate(currentUser, today);
        log.info("Found {} workout plans for today", plans.size());
        return plans;
    }

    @Transactional(readOnly = true)
    public List<WorkoutPlan> getUserWorkoutPlans() {
        User currentUser = userService.getCurrentUser();
        return workoutPlanRepository.findByUser(currentUser);
    }

    @Transactional(readOnly = true)
    public List<WorkoutPlan> getActiveWorkoutPlans() {
        User currentUser = userService.getCurrentUser();
        return workoutPlanRepository.findByUserAndActiveStatus(currentUser, WorkoutPlan.ActiveStatus.ACTIVE);
    }

    @Transactional(readOnly = true)
    public WorkoutPlan getWorkoutPlanById(Long id) {
        WorkoutPlan workoutPlan = workoutPlanRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Workout plan not found with id: " + id));
        
        User currentUser = userService.getCurrentUser();
        if (!workoutPlan.getUser().getId().equals(currentUser.getId())) {
            throw new RuntimeException("Access denied: You can only access your own workout Plans");
        }
        
        return workoutPlan;
    }

    @Transactional
    public WorkoutPlan updateWorkoutPlan(Long id, WorkoutPlan workoutPlanDetails) {
        WorkoutPlan workoutPlan = getWorkoutPlanById(id);
        
        workoutPlan.setTitle(workoutPlanDetails.getTitle());
        workoutPlan.setDescription(workoutPlanDetails.getDescription());
        workoutPlan.setActiveStatus(workoutPlanDetails.getActiveStatus());
        
        return workoutPlanRepository.save(workoutPlan);
    }

    @Transactional
    public void deleteWorkoutPlan(Long id) {
        WorkoutPlan workoutPlan = getWorkoutPlanById(id);
        workoutPlanRepository.delete(workoutPlan);
    }

    @Transactional
    public WorkoutPlan completeWorkoutPlan(Long id) {
        WorkoutPlan workoutPlan = getWorkoutPlanById(id);
        workoutPlan.setActiveStatus(WorkoutPlan.ActiveStatus.COMPLETED);
        log.info("Workout plan {} marked as completed", id);
        return workoutPlanRepository.save(workoutPlan);
    }

    @Transactional
    public int autoCompleteExpiredWorkouts() {
        try {
            User currentUser = userService.getCurrentUser();
            java.time.LocalDate today = java.time.LocalDate.now();
            java.time.LocalTime currentTime = java.time.LocalTime.now();
            
            log.info("Auto-completing workouts for user {} at {} on {}", 
                    currentUser.getEmail(), currentTime, today);
            
            // Find all active workouts that have passed their end time
            java.util.List<WorkoutPlan> expiredWorkouts = workoutPlanRepository
                    .findByUserAndActiveStatusAndWorkoutDateLessThanEqual(currentUser, WorkoutPlan.ActiveStatus.ACTIVE, today);
            
            log.info("Found {} active workouts to check for auto-completion", expiredWorkouts.size());
            
            int completedCount = 0;
            for (WorkoutPlan workout : expiredWorkouts) {
                // Check if workout date is in the past OR (today AND end time has passed)
                if (workout.getWorkoutDate().isBefore(today) || 
                    (workout.getWorkoutDate().equals(today) && workout.getEndTime().isBefore(currentTime))) {
                    workout.setActiveStatus(WorkoutPlan.ActiveStatus.COMPLETED);
                    workoutPlanRepository.save(workout);
                    completedCount++;
                    log.info("Auto-completed workout plan {}", workout.getId());
                }
            }
            
            if (completedCount > 0) {
                log.info("Auto-completed {} expired workout(s)", completedCount);
            }
            
            return completedCount;
        } catch (Exception e) {
            log.error("Error in autoCompleteExpiredWorkouts: {}", e.getMessage(), e);
            throw e;
        }
    }
}

File: frontend/src/services/workoutService.ts
const API_BASE_URL = `${import.meta.env.VITE_API_BASE_URL || 'http://localhost:8080/api'}/v1`;

// Transform backend exercise types to user-friendly labels
const transformExerciseTypes = (types: string[]): ExerciseType[] => {
  const typeLabels: Record<string, string> = {
    'AEROBIC': 'Aerobic',
    'STRENGTH': 'Strength Training', 
    'FLEXIBILITY': 'Flexibility',
    'BALANCE': 'Balance'
  };
  
  return types.map(type => ({
    value: type as 'AEROBIC' | 'STRENGTH' | 'FLEXIBILITY' | 'BALANCE',
    label: typeLabels[type] || type
  }));
};

export interface WorkoutPlanCreateRequest {
  exerciseType: string;
  exerciseId: number;
  workoutDate: string;
  startTime: string;
  endTime: string;
  comments?: string;
  workoutPhotoUrl?: string;
  title?: string;
  description?: string;
}

export interface ExerciseType {
  value: 'AEROBIC' | 'STRENGTH' | 'FLEXIBILITY' | 'BALANCE';
  label: string;
}

export interface Exercise {
  id: string;
  name: string;
  description: string;
  category: string;
}

export interface WorkoutPlan {
  id: number;
  title: string;
  description?: string;
  exerciseType: string;
  exercise: Exercise;
  workoutDate: string;
  startTime: string;
  endTime: string;
  comments?: string;
  workoutPhotoUrl?: string;
  activeStatus: string;
  createdAt: string;
  updatedAt: string;
}

class WorkoutService {
  private baseUrl: string;

  constructor() {
    this.baseUrl = API_BASE_URL;
  }

  private getAuthHeaders(): HeadersInit {
    const token = localStorage.getItem('token');
    return {
      'Content-Type': 'application/json',
      ...(token && { 'Authorization': `Bearer ${token}` })
    };
  }

  async getExerciseTypes(): Promise<ExerciseType[]> {
    try {
      const response = await fetch(`${this.baseUrl}/workouts/exercise-types`, {
        headers: this.getAuthHeaders()
      });
      if (!response.ok) {
        throw new Error('Failed to fetch exercise types');
      }
      const types: string[] = await response.json();
      return transformExerciseTypes(types);
    } catch (error) {
      console.warn('Backend exercise types fetch failed, using mock data');
      // Fallback to mock exercise types
      return transformExerciseTypes(['AEROBIC', 'STRENGTH', 'FLEXIBILITY', 'BALANCE']);
    }
  }

  async getExercisesByCategory(category: string): Promise<Exercise[]> {
    if (!category) {
      return [];
    }
    
    try {
      const response = await fetch(`${this.baseUrl}/workouts/exercises/by-category?category=${category}`, {
        headers: this.getAuthHeaders()
      });
      if (!response.ok) {
        throw new Error(`Failed to fetch exercises for category: ${category}`);
      }
      return response.json();
    } catch (error) {
      console.warn(`Backend exercises fetch failed for category ${category}, using mock data`);
      // Fallback to mock exercises - use ExerciseService for consistency
      const ExerciseService = (await import('./exerciseService')).default;
      return ExerciseService.getExercisesByCategory(category.toLowerCase() as any);
    }
  }

  async createWorkoutPlan(workoutData: WorkoutPlanCreateRequest): Promise<WorkoutPlan> {
    // Transform data for backend - add seconds to time for LocalTime parsing
    const backendData = {
      ...workoutData,
      startTime: workoutData.startTime ? `${workoutData.startTime}:00` : workoutData.startTime,
      endTime: workoutData.endTime ? `${workoutData.endTime}:00` : workoutData.endTime,
    };
    
    const response = await fetch(`${this.baseUrl}/workouts`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'Authorization': `Bearer ${localStorage.getItem('token') || ''}`
      },
      body: JSON.stringify(backendData)
    });

    if (!response.ok) {
      throw new Error('Failed to create workout plan');
    }
    return response.json();
  }

  async getTodayWorkoutPlans(): Promise<WorkoutPlan[]> {
    try {
      const response = await fetch(`${this.baseUrl}/workouts/today`, {
        headers: this.getAuthHeaders()
      });

      if (!response.ok) {
        throw new Error('Failed to fetch today\'s workout plans');
      }
      return response.json();
    } catch (error) {
      console.warn('Backend today workouts fetch failed, using empty array');
      // Fallback to empty array for today's workouts
      return [];
    }
  }

  async getAllWorkoutPlans(): Promise<WorkoutPlan[]> {
    const response = await fetch(`${this.baseUrl}/workouts`, {
      headers: {
        'Authorization': `Bearer ${localStorage.getItem('token') || ''}`
      }
    });

    if (!response.ok) {
      throw new Error('Failed to fetch workout plans');
    }
    return response.json();
  }

  async getUserWorkoutPlans(): Promise<WorkoutPlan[]> {
    const response = await fetch(`${this.baseUrl}/workouts`, {
      headers: this.getAuthHeaders()
    });

    if (!response.ok) {
      throw new Error('Failed to fetch user workout plans');
    }
    return response.json();
  }

  async updateWorkoutPlan(id: number | string, workoutData: Partial<WorkoutPlanCreateRequest>): Promise<WorkoutPlan> {
    const token = localStorage.getItem('token');
    if (!token) {
      throw new Error('No authentication token found. Please log in.');
    }

    // Transform to backend format
    const backendData: any = {
      ...workoutData,
    };
    
    if (workoutData.exerciseId) {
      backendData.exerciseId = workoutData.exerciseId;
    }
    if (workoutData.workoutDate) {
      backendData.workoutDate = workoutData.workoutDate;
    }
    if (workoutData.startTime) {
      backendData.startTime = workoutData.startTime + ':00';
    }
    if (workoutData.endTime) {
      backendData.endTime = workoutData.endTime + ':00';
    }

    const response = await fetch(`${this.baseUrl}/workouts/${id}`, {
      method: 'PUT',
      headers: {
        'Content-Type': 'application/json',
        'Authorization': `Bearer ${token}`
      },
      body: JSON.stringify(backendData)
    });

    if (!response.ok) {
      const errorText = await response.text();
      throw new Error(`Failed to update workout plan: ${errorText}`);
    }
    return response.json();
  }

  async completeWorkoutPlan(id: number | string): Promise<WorkoutPlan> {
    const token = localStorage.getItem('token');
    if (!token) {
      throw new Error('No authentication token found. Please log in.');
    }

    const response = await fetch(`${this.baseUrl}/workouts/${id}/complete`, {
      method: 'POST',
      headers: {
        'Authorization': `Bearer ${token}`
      }
    });

    if (!response.ok) {
      throw new Error('Failed to complete workout plan');
    }
    return response.json();
  }

  async autoCompleteExpiredWorkouts(): Promise<number> {
    const token = localStorage.getItem('token');
    if (!token) {
      throw new Error('No authentication token found. Please log in.');
    }

    const response = await fetch(`${this.baseUrl}/workouts/auto-complete`, {
      method: 'POST',
      headers: {
        'Authorization': `Bearer ${token}`
      }
    });

    if (!response.ok) {
      throw new Error('Failed to auto-complete workouts');
    }
    return response.json();
  }

  async deleteWorkoutPlan(id: number | string): Promise<void> {
    const token = localStorage.getItem('token');
    if (!token) {
      throw new Error('No authentication token found. Please log in.');
    }

    const response = await fetch(`${this.baseUrl}/workouts/${id}`, {
      method: 'DELETE',
      headers: {
        'Authorization': `Bearer ${token}`
      }
    });

    if (!response.ok) {
      throw new Error('Failed to delete workout plan');
    }
  }
}

export default new WorkoutService();
